presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/StochasticPriorityQueue.java

```diff
@@ -0,0 +1,261 @@
+                Node<E> leaf = root.findLeaf();
+                leaf.remove();
+                node.setTickets(leaf.getTickets());
+                node.setValue(leaf.getValue());
+                index.put(leaf.getValue(), node);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public E poll()
+    {
+        if (root == null) {
+            return null;
+        }
+
+        long winningTicket = ThreadLocalRandom.current().nextLong(root.getTotalTickets());
+        Node<E> candidate = root;
+        while (!candidate.isLeaf()) {
+            long leftTickets = candidate.getLeft().map(Node::getTotalTickets).orElse(0L);
+
+            if (winningTicket < leftTickets) {
+                candidate = candidate.getLeft().get();
+                continue;
+            }
+            else {
+                winningTicket -= leftTickets;
+            }
+
+            if (winningTicket < candidate.getTickets()) {
+                break;
+            }
+            else {
+                winningTicket -= candidate.getTickets();
+            }
+
+            checkState(candidate.getRight().isPresent(), "Expected right node to contain the winner, but it does not exist");
+            candidate = candidate.getRight().get();
+        }
+        checkState(winningTicket < candidate.getTickets(), "Inconsistent winner");
+
+        E value = candidate.getValue();
+        remove(value);
+        return value;
+    }
+
+    @Override
+    public int size()
+    {
+        return index.size();
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return index.isEmpty();
+    }
+
+    private static final class Node<E>
+    {
+        private Optional<Node<E>> parent;
+        private E value;
+        private Optional<Node<E>> left = Optional.empty();
+        private Optional<Node<E>> right = Optional.empty();
+        private int tickets;
+        private long totalTickets;
+        private int decendants;
+
+        private Node(Optional<Node<E>> parent, E value)
+        {
+            this.parent = parent;
+            this.value = value;
+        }
+
+        public E getValue()
+        {
+            return value;
+        }
+
+        public void setValue(E value)
+        {
+            this.value = value;
+        }
+
+        public Optional<Node<E>> getLeft()
+        {
+            return left;
+        }
+
+        public Optional<Node<E>> getRight()
+        {
+            return right;
+        }
+
+        public long getTotalTickets()
+        {
+            return totalTickets;
+        }
+
+        public int getTickets()
+        {
+            return tickets;
+        }
+
+        public void setTickets(int tickets)
+        {
+            checkArgument(tickets > 0, "tickets must be positive");
+            Node<E> node = this;
+            while (node != null) {
+                node.totalTickets -= this.tickets;
+                node.totalTickets += tickets;
```

Comment:
I could be wrong, but this seems broken.  It looks like we are overwriting the parent nodes with this node's total